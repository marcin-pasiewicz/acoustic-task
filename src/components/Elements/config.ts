export const elementsOrder = ["heading", "mainImage", "body", "author", "date"];
