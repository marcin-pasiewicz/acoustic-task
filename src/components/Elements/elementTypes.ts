export enum ElemeneTypes {
  author = "author",
  body = "body",
  date = "date",
  heading = "heading",
  mainImage = "mainImage",
}
