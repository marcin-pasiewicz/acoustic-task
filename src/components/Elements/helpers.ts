export function prepareElements(elements: any, order: any[]) {
  return order.map(item => {
    return elements[item] && {
      ...elements[item],
      elementName: item,
    }
  }).filter(item => item !== undefined)
}