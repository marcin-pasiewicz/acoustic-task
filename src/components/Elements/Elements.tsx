import React from "react";
import styled from "styled-components";
import { Element } from "../../models/Content";
import { ElemeneTypes as ElemeneNames } from "./elementTypes";
import Heading from "../Heading/Heading";
import Author from "../Author/Author";
import MainImage from "../MainImage/MainImage";
import Body from "../Body/Body";
import Date from "../Date/Date";
import { elementsOrder } from "./config";
import { prepareElements } from "./helpers";
import { isEmpty } from "../../helpers/objectHelpers";

interface Props {
  elements: {
    [key: string]: Element;
  };
}

export default function Elements({ elements }: Props) {
  const processedElements = prepareElements(elements, elementsOrder);

  function getElement(element: any) {
    switch (element.elementName) {
      case ElemeneNames.heading:
        return <Heading {...element} />;
      case ElemeneNames.mainImage:
        return <MainImage {...element} />;
      case ElemeneNames.body:
        return <Body {...element} />;
      case ElemeneNames.author:
        return <Author {...element} />;
      case ElemeneNames.date:
        return <Date {...element} />;
      default:
        return <div>Unknown element</div>;
    }
  }
  return (
    <ElementsWrapper>
      {!isEmpty(elements) &&
        processedElements.map((element, index) => (
          <div key={index}>{getElement(element)}</div>
        ))}
    </ElementsWrapper>
  );
}

const ElementsWrapper = styled.div`
  max-width: 600px;
  margin: auto;
`;
