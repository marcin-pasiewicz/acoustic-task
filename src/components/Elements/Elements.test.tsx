import React from "react";
import { render } from "@testing-library/react";
import Elements from "./Elements";

const testData = {
  elements: {
    heading: { elementType: "text", value: "Staffordshire Terrier Pups" },
  },
};

describe("<Elements/>", () => {
  it("renders correctly", () => {
    const { asFragment } = render(<Elements elements={testData} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
