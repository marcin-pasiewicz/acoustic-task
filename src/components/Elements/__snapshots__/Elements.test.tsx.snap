// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Elements/> renders correctly 1`] = `
<DocumentFragment>
  <div
    class="sc-AxhCb dTHeSs"
  />
</DocumentFragment>
`;
